content/blog/2023-08-03-parsing-ssh-config.markdown

```diff
@@ -0,0 +1,210 @@
+the config.
+
+1. For starters, the config is parsed line by line. Leading whitespace (i.e.,
+   indentation) is ignored. So, while indentation makes it look like you are
+   configuring properties for a particular host, this isn't quite correct.
+   Instead, the `Host` and `Match` lines are special statements that enable or
+   disable all subsequent lines until the next `Host` or `Match`.
+
+    There is no backtracking; previous conditions and lines are not re-evaluated
+    after learning more about the config later on.
+
+2. When a config line is seen, and is active thanks to the most recent `Host` or
+   `Match` succeeding, its value is selected if it is the first of that config
+   to be selected. So the earliest place a value is set takes priority; this may
+   be a little counterintuitive if you are used to having the latest value be
+   picked, such as enable/disable command line flags tend to work.
+
+3. When `HostName` is set, it replaces the value used in `Host` matches and the
+   `host` value in `Match` matches.
+
+4. The last behaviour of interest is the `Match final` rule. There are several
+   conditions a `Match` statement can have, and the `final` rule says make this
+   active on the final pass over the config.
+
+Wait, final pass? Multiple passes? Yes. If `final` is a condition on a `Match`,
+SSH will do another pass over the config, following all the rules above. Except
+this time all the configs we read on the first pass are still active (and can't
+be changed). But all the `Host` and `Matches` are re-evaluated, allowing other
+configs to potentially be set. I guess that means rule (1) ought to have a big
+asterisk next to it.
+
+Together, these rules can lead to some quirky behaviours. Consider the following
+config
+
+```text
+Host *.ozlabs.ibm.com
+    ProxyJump proxy
+
+Host example
+    HostName example.ozlabs.ibm.com
+```
+
+If I run `ssh example` on the command line, will it use the proxy?
+
+By rule (1), no. When testing the first `Host` condition, our host value is
+currently `example`. It is not until we reach the `HostName` config that we
+start using `example.ozlabs.ibm.com` for any host matches.
+
+But by rule (4), the answer turns into _maybe_. If we end up doing a second pass
+over the config thanks to a `Match final` that could be _anywhere_ else, we
+would now be matching `example.ozlabs.ibm.com` against the first `Host` on the
+second go around. This will pass, and, since nothing has set `ProxyJump` yet, we
+would gain the proxy.
+
+You may think, yes, but we don't have a `Match final` in that example. But if
+you thought that, then you forgot about the system config.
+
+The system config is effectively appended to the user config, to allow any
+system wide settings. Most of the time this isn't an issue because of the
+first-come-first-served rule with config matches (rule 2). But if the system
+config includes a `Match final`, it will trigger the entire config to be
```

Comment:
I would mention that on Fedora that's in /etc/ssh/ssh_config.d, so this is a real thing that actually happens

Reply:
I don't think it's necessary I just think things like that help connect concepts to what someone is running in reality

Reply:
I agree, you should talk about the concrete example and where the file is